### PR TITLE
Stateful Client

### PR DESCRIPTION
**Pull request by jen20**
Further to the discussion #27 this pull request converts the Azure SDK to having a stateful client, allowing multiple subscriptions to be operated on in one process. It also makes the package name compliant with the go standards (removing the need to alias it) and flattens the hierarchy of packages from one-per-client down to just one.

Given the size of this refactor I fully expect there to be discussion etc around whether this will be accepted into this project or whether the API changes are too vast and it needs to become something different. As far as I can tell the tests all pass, though ultimately Travis will be the arbiter of that!

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
cc: @paulmey @hopetobelievein @jeffmendoza

File: azure/errors.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package azure
+
+const (
+	publishSettingsConfigurationError = "PublishSettingsFilePath is set. Consequently ManagementCertificatePath and SubscriptionId must not be set."
+
+	managementCertificateConfigurationError = "ManagementCertificatePath is set. Consequently SubscriptionId must also be set, and PublishSettingsFilePath must not be set."
```
Comment (hopetobelievein) [inline azure/errors.go]: How do you know that ManagementCertificatePath  is set?
Condition was: len(config.ManagementCertificatePath) == 0 || len(config.SubscriptionId) == 0
They both can be null at this point.
Mb we should say something like "Both ManagementCertificatePath and SubscriptionId should be set, and PublishSettingsFilePath must not be set."

File: azure/errors.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package azure
+
+const (
+	publishSettingsConfigurationError = "PublishSettingsFilePath is set. Consequently ManagementCertificatePath and SubscriptionId must not be set."
```
Comment (hopetobelievein) [inline azure/errors.go]: Looks like publishSettingsConfigurationError and managementCertificateConfigurationError constants are specific for client.go file. And all of other file-specific constants are stored in their own files. 
I think we should use this file for reusable consts only, or move everything here.

**Comment by ahmetalpbalkan** [comment]
@jen20 thanks, this is huge! This'll probably take some time to review. We should probably release this as 2.0 as this'll break all projects depending on clients.

We should keep different clients in different packages. For example in docker-registry project I am just using blob storage client and I don't want to download the entire thing. Any build break or serious bug in one client will break others very easily.

See azure-sdk-python or azure-sdk-for-net for examples. They separate different REST APIs into different projects. It provides a great deal of modularity. Let's keep `clients/[clientname]` path naming and packages separate, shall we?

File: azure/errors.go (hunk 1 above)
Comment (ahmetalpbalkan) [inline azure/errors.go]: also errors must be named as `ErrXxx` or `errXxx` (see go stdlib for convention)

**Comment by hopetobelievein** [comment]
@ahmetalpbalkan +1 for package separation.

File: azure/errors.go (hunk 1 above)
Comment (jen20) [inline azure/errors.go]: This needs more attention, it's completely untested with anything except a publish settings file. Also it needs a better way to choose a subscription from a file with more than one.

**Comment by jen20** [comment]
The problem there is that client has to import them all to provide access, or they can't be added to the type. Open to suggestions here though.

**Comment by ahmetalpbalkan** [comment]
@jen20 You can introduce a composition like this: There's package `common` or `azure` where credentials live. From there if I import `clients/virtualnetwork`, I initialize `VirtualNetworkClient` with `Credentials` object I got from the `azure` (or `common`) package. That way you don't introduce closely coupled packages or circular dependencies. How's that sound?

**Comment by jeffmendoza** [comment]
Should we pull this into a branch, to collaborate on?

**Comment by ahmetalpbalkan** [comment]
@jeffmendoza agreed. once we're done reviewing this, we should probably merge it to some branch other than `master`. objections?

**Comment by jen20** [comment]
@ahmetalpbalkan that could work. Not sure I'd agree with keeping things other than the latest stuff on master, it would be better to tag releases? Admittedly this will break people in the short term. If not it might be better to make it a completely separate set of packages in a different repo such that `go get` works nicely with this version?

**Comment by ahmetalpbalkan** [comment]
@jen20 idea around working on it in a separate branch is because most likely we'll need to make it work better and polish the code a bit more after this PR. We'll be eventually merging to `master` for sure. :wink:

**Comment by jen20** [comment]
Have put a few commits in to deal with some of the issues raised above. I'll take a shot now at separating the packages back out.

**Comment by jen20** [comment]
I'm following the convention that error names have followed throughout this codebase for now - if we want to change them later we can do, but at the moment they're all `XxxError`.

**Comment by ahmetalpbalkan** [comment]
@jen20  ...except the `storage` client. :smile: https://github.com/MSOpenTech/azure-sdk-for-go/blob/master/clients/storage/blob.go#L255 we do want to change them and this sounds like a good time. if you don't mind changing that'd be awesome.

**Comment by jen20** [comment]
Will take a look at constant naming in a bit, trying to get the main structural changes in first.

**Comment by jen20** [comment]
OK cool looks like travis is sorted again. Any comments on the structural changes before I go through and do style fixes?

**Comment by jen20** [comment]
Hah, forgot about `:GoRename` - changing these was actually fairly straightforward.

File: azure/imageclient/client.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	errParamNotSpecified = "Parameter %s is not specified."
+)
+
+//ImageClient is used to manage operations on Azure Locations
+type ImageClient struct {
+	client *azure.Client
+}
+
+//NewClient is used to instantiate a new ImageClient from an Azure client
+func NewClient(client *azure.Client) *ImageClient {
```
Comment (hopetobelievein) [inline azure/imageclient/client.go]: What about moving all client sructs like ImageClient and HostedServiceClient to entities files? This way all classes will be in one place.

File: azure/imageclient/client.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+const (
+	azureImageListURL    = "services/images"
+	errInvalidImage      = "Can not find image %s in specified subscription, please specify another image name."
+	errParamNotSpecified = "Parameter %s is not specified."
+)
+
+//ImageClient is used to manage operations on Azure Locations
+type ImageClient struct {
```
Comment (hopetobelievein) [inline azure/imageclient/client.go]: What about moving all client sructs like ImageClient and HostedServiceClient to entities files? This way all classes will be in one place.

Comment (ahmetalpbalkan) [inline azure/imageclient/client.go]: doesn't really add much value IMHO.

**Comment by ahmetalpbalkan** [comment]
@jen20 feel free to send small commits all the time, those are much easier to review, or for the big changes, feel free to squash at your judgement.

Btw I think we should rename our pkg names while we're at it e.g.

* `clients/hostedServiceClient` to `clients/hostedservice`
* `clients/vmDiskClient` to `clients/vmdisk`

sounds like more precise and to the point. I find these useful:
https://talks.golang.org/2014/organizeio.slide#6
https://golang.org/doc/effective_go.html#package-names

**Comment by jen20** [comment]
@ahmetalpbalkan As per http://blog.golang.org/package-names, they shouldn't contain mixed case (I changed them from `hostedServiceClient` to `hostedserviceclient` etc. However, `hostedservice`, `location` etc also comes under "don't steal good names" - `location` for example is the perfect variable name when this is in use in application code and I think we should avoid conflicting with it.

File: azure/imageclient/client.go (hunk 4 above)
Comment (jen20) [inline azure/imageclient/client.go]: Moving the struct is fine, I'd keep the `NewClient()` in the main client file in each case though as it makes it more discoverable when browsing the code.

**Comment by ahmetalpbalkan** [comment]
@jen20 I unfortunately hostedservice is better than hostedserviceclient. :) It may look like a perfect variable name but then golang encourages using short variable names e.g. `l` or `loc` for `location` (see stdlib for more). I just peeked at [gcloud-golang](https://github.com/GoogleCloudPlatform/gcloud-golang) for the first time and you can argue the same thing about that as well, but not suffixing with “client” will make it precise and readable.

**Comment by jen20** [comment]
I can see that one working either way. In my case I'd just alias them all, since `l` is a properly crap name for a variable denoting what region you're working in when you're actually using this stuff in something more complex than a unit test. What about when we come to `vm`?

Actually it might be worth bringing some of the common ones back into the root namespace. Most things will rely on locations etc.

**Comment by ahmetalpbalkan** [comment]
> What about when we come to vm?

not sure what you're referring to.

Actually one wrong approach I think we took is having logic in this library. There are certain methods that do more than one REST call and implements a common sequence of operations. This is not the practice we usually follow in other language SDKs we have. If you're referring to the fact that vmClient uses locationClient, there's not much about we can do about it.

I'd say lets shorten those names and let users alias.

**Comment by jen20** [comment]
I mean we'll end up with a package named `vm`, which again is a perfect variable name to grab the config. I'd kind of agree that putting logic in here was a bad plan overall - perhaps it would be better to drop down to just implementing the various API calls and allow other libraries to do the orchestration?

**Comment by ahmetalpbalkan** [comment]
`vm` → `virtualmachine`. problem solved. :smile: I'm not really concerned about package names, those can be easily aliased.

**Comment by jen20** [comment]
That seems reasonable enough while they can be aliased. I'll go through and change them.

**Comment by jen20** [comment]
Heh, 231e554 is important... GoImports by default imported all the default HTTP and TLS stuff instead of the forks in this project. Might be worth giving #34 some consideration as this would have broken everyone despite Travis reporting everything is OK.

**Comment by ahmetalpbalkan** [comment]
@jen20 agreed. we can write unit tests around request methods by mocking the Azure Service Management REST API in an http.Server, however I'm not sure if it's worth the trouble because it'll be just testing these imports and few other small things and there will be things that we will forgot to mock from REST API. So things passing in unit tests will actually fail in reality.

You can check how the Python SDK implements tests. https://github.com/Azure/azure-sdk-for-python/tree/master/tests Normally we have a storage emulator for Windows but we don't have an emulator for Service Management REST APIs. Therefore I'm not sure how we can have CI tests around service management operations without actually hitting the API.

cc: @huguesv for providing insights on how Python SDK is tested.

**Comment by jen20** [comment]
Could we just hit the API for integration tests? Agreed, mocking the API seems like a bad plan overall (it will never match correctly). Normally the consideration here would be the cost, but can MS do something internally?

**Comment by jeffmendoza** [comment]
I've checked in the code here https://github.com/MSOpenTech/azure-sdk-for-go/tree/jen20-stateful-client and added @jen20 as a collaborator. This way we can work together in that branch until the code is stable and we have a transition plan. How does this sound?

**Comment by ahmetalpbalkan** [comment]
@jeffmendoza sounds good to me. I didn't have a chance to review the entire change. I hope somebody else has gone through it. :)

**Comment by jen20** [comment]
@jeffmendoza I'll close this PR in favour of one from the branch in this repository.

**Comment by hopetobelievein** [comment]
@ahmetalpbalkan @jeffmendoza I've seen it all, looks good in general.

**Pull request closed without merging**